### PR TITLE
Flyout and Footer API design document

### LINKED ISSUE #7748: Make our documentation page APIs more user friendly
This issue is a bit of a meta issue, but it also has at least one concrete action associated with it.

### The high level

We want to make it easier for theme authors to integrate the information in our flyout into their themes, which means giving hooks into the information. The most commonly used data here is:

* Active versions
* Translations
* Downloads (PDF, HTML, Epub)
* View/Edit on GitHub links

### Short term work

The most obvious place to focus in the short term is breaking out our monolithic "footer API" into something more usable by users. Currently this API returns rendered HTML with the data instead, along with a few additional JSON details. It doesn't include in JSON the actual data used to render the HTML.

So I imagine the steps towards transition are:

* Add additional JSON data to the current response
* Migrate our code to use this new JSON data structure
* Remove the existing HTML data structure after a deprecation period

I don't believe many users are currently using this structure, but I do know some people have custom CSS on top of it, so we should try and not alter the final HTML output if we can avoid it. 

##### Current response

Screen Shot 2020-12-14 at 9 10 27 AM

##### Current flyout

Screen Shot 2020-12-14 at 9 12 43 AM

### Longer term vision

Once we are returning the JSON data, we need to find a pattern for theme authors to "hook" into this data with Javascript. I think there are a few different approaches to this, and I don't know what the best option would be. I just want to note that is probably the "v2" after we migrate to using the JSON data approach is actually exposing that data to users in a standard way.

Refs #7732

### PR DESCRIPTION
**Pull request by davidfischer**
This is a design document for using APIv3 in order to generate the flyout (version selector) menu. This is purely a proposal at this point to get everyone on the same page with respect to goals and how it could be implemented.

The main goals are to:

* Give themes and projects a way to control the flyout menu. More than just CSS styling, they could take the raw data and put it into the navigation or other menus. For example, this would allow building Django's version selector menu instead of using RTD's default. A custom Sphinx or MkDocs theme could take on this work or a project could override it themselves.
* Use the new APIv3 endpoint instead of an undocumented APIv2 endpoint that we're trying to deprecate.
* Provide a baseline for overriding various other aspects of how Read the Docs customizes documentation.

For easier reading, there's a [link to the document](https://docs--8052.org.readthedocs.build/en/8052/development/design/flyout-redesign.html).

Closes #7748

### CODE REVIEW

File: docs/development/design/flyout-redesign.rst
Comment (stsewd) [inline docs/development/design/flyout-redesign.rst:130]: we already do this, but we haven't document them. I like more the name you are proposing here p:

`READTHEDOCS_DATA.features.docsearch_disabled`

File: docs/development/design/flyout-redesign.rst
Comment (stsewd) [inline docs/development/design/flyout-redesign.rst:122]: I'd name this custom_version_selector maybe? Wasn't obvious for me that this was a custom function to inject the footer.

File: docs/development/design/flyout-redesign.rst
Comment (stsewd) [inline docs/development/design/flyout-redesign.rst:49]: Just a note here that we are using the proxied endpoint of the footer, this `/_/api/v2/footer_html`, this allow us to share the same domain with some advantages:

- avoid being blocked by some extensions
- Allow to cache the response (docs are served from cloudflare)
- Allow to use the auth backends in .com

And, I've been having some ideas around our APIs... here is an extract from https://github.com/readthedocs/readthedocs.org/pull/8039/

>    We should probably make a distinction between our general API that handles Read the Docs resources,
   vs our APIs that expose features (like server side search, footer, and embed, all of them proxied).
   This way we can version each endpoint separately.

I do agree that the footer shares a lot with the projects/ endpoint. Proxying api v3 may be a little of a problem, because it relies heavily on resolving URLs from the dashboard, but in proxito we don't have the url patterns from the main site.

File: docs/development/design/flyout-redesign.rst
Comment (davidfischer) [reply docs/development/design/flyout-redesign.rst:130]: I know we do this. I like the idea a lot but I wish it was easier for users to control as well as documented. The way it is now can run into race conditions with us defining these features and users setting them.

File: docs/development/design/flyout-redesign.rst
Comment (davidfischer) [reply docs/development/design/flyout-redesign.rst:49]: This is a very good point. Maybe we don't need all the APIv3 endpoints but we can have an endpoint that just returns the same data as the APIv3 endpoint and shares code.

File: docs/development/design/flyout-redesign.rst
Comment (humitos) [inline docs/development/design/flyout-redesign.rst:55]:
```suggestion
These options already partially implemented via the `DRF Flex Fields <https://pypi.org/project/drf-flex-fields/>`_ module
```

File: docs/development/design/flyout-redesign.rst
Comment (humitos) [inline docs/development/design/flyout-redesign.rst:58]: It seems that we could implement `?expand=translations,subrpojects` and we would have everything we need with one API call: https://readthedocs.org/api/v3/projects/docs/?expand=active_versions,translations,subprojects

One downside that has using this endpoint is that we will be returning much data than we need and that could slow down the response time.

Another thing to consider is that APIv3 is authenticated-only, and we require anonymous access for the flyout.

Comment (humitos) [reply docs/development/design/flyout-redesign.rst:58]: I'm not sold on the idea of sharing the endpoint for the flyout than a regular endpoint. In the future we may want to _add extra data_ to `/api/v3/projects/` and we may be blocked because it increases the response time by some milliseconds that are not acceptable for the flyout endpoint.

I think that having an independent endpoint with more control over it is a win here and will avoid headaches in the future.

**Review by humitos — COMMENTED** [review]
The proposal looks good to me. I don't have too many comments on the JS implementation side, but seems it will work to me.

I'm not convinced that it's a good idea to share the APIv3 endpoint for the flyout to a regular endpoint. I think having independence and granular control here is better.

**Review by ericholscher — COMMENTED** [review]
This is a good start. I think we probably need to clarify our names a good bit here, but I like the general approach here. I don't know enough about JS to really judge alternative implementations though. I know our ad client works in a much more complicated way, but I'm not sure if that's because it's used on more sites, or why that design was chosen. 

In particular, I think we need to clarify and standardize on names for:

* What this API is called. `footer` isn't really correct. I think probably it's `flyout_data` or something like that? `dynamic`project_data`?
* What the flyout is called? I think we generally standardized on `flyout` but this doc uses `version_selector`. Should we remove `flyout` for the various composite parts?

I'm sure there's more named needed, but those were the two obvious ones. Names are hard.

**Review by agjohnson — COMMENTED** [review]
I'm still considering the effects of using global state to configure the client. Generally, having a fully instantiated object is a more expected pattern and an easier pattern to work with if the user wants to extend functionality instead of replace it.

**Comment by humitos** [comment]
Things have changed considerably here with the introduction of the addons in https://github.com/readthedocs/readthedocs-client. I think it's safe to close this PR or, if we want, merge and close so we don't loose the document. Maybe adding a note in the top of it linking to the addons repository?

**Comment by agjohnson** [comment]
Maybe we just keep this available for reference on this branch? Merging this could be confusing for anyone that comes across this first, as we deviated from this plan a bit. I agree adding a note would help either way though :+1:

**Comment by humitos** [comment]
I updated the warning note and I will merge it after that.

**Review by humitos — APPROVED** [review]
I'm approving this PR so it gets merged automatically after the check pass. Note that this work is _not implemented_ and most of it was replaced by https://github.com/readthedocs/readthedocs-client and its new endpoint (`/_/readthedocs-config/`)

**Pull request merged**
